posts/centralized-rules-execution-server.md

```diff
@@ -0,0 +1,100 @@
+At one of our recent customer's location, we were tasked to resolve a problem and the application has to look into more than hundred different scenarios to come up with a result. Also the customer wanted to have a provision to maintain the rules separately as the rules can change quite often so that the main applications do not have to be redployed whenever there is a rule change.
+
+## The solution
+We decided to use a centralized rules execution engine based on Jboss Drools (Kie execution server) to make the rules engine a loosely coupled application, to maintain the rules outside of the business application, to make the rules engine available to other applications in the organization which have similar use cases.
+
+Kie execution server can be used to instantiate and execute rules via Rest, JMS or Java client side applications, it supports runtime updates to the rules.
+
+
+## Architecture
+The first version of the rules engine is designed to adhere to the customer's infrastructure, ability to use the customer's source control repository, security standards and CI/CD process. The solution is designed to auto-recover in case of rules engine restarts/crashes.
+
+- ### Deployment architecture
+The rules engine is deployed into the existing AWS ECS infrastructure as a spring boot application. The service has access to AWS EFS service so that it can store its state in the metadata file on the network so that it can recover from the unexpected restarts.
+
+The service has access to the internal maven repository so that it can access the rules artifacts and load them as Kie Containers. Each Kie Container represents one set of rules maintained by a project team.
+
+The service has access to customer's LDAP to support security and authorizations, so that only the authorized users have access to update the rules.
+
+![alt](https://raw.githubusercontent.com/msambaraju/blog-usa/master/images/2019/01/Kie_Server.png)
+
+- ### Development architecture
+Each development team is responsible for the maintenance of their rules, the rules are maintained in the git repository as maven projects. The project contains the rules files in .drl or .xls formats and any dependencies are maintained in the pom.xml such as POJO's used in the rules, or any other utilities used in the rules files.
+
+The project is configured to built and packaged using the existing Jenkins pipeline into a jar file. The built artifact is pushed into the internal maven repository and it complies to the existing release management process.
+
+![alt](https://raw.githubusercontent.com/msambaraju/blog-usa/master/images/2019/01/Kie_Container_Process.png)
+
+## Implementation
+The base spring boot application is generated by JHipster so that the capabilities provided by JHipster are already part of the generated application. For more information on JHipster visit [JHipster](https://www.jhipster.tech). JHipster is already standardized at the customer location to support existing infrastructure.
```

Comment:
```suggestion
The base spring boot application is generated by [JHipster](https://www.jhipster.tech) so that the out-of-the-box capabilities provided by JHipster are already part of the generated application. JHipster has become a standard application at the customer location to support existing infrastructure.
```